lib/Auth/Process/ExpiryDate.php

```diff
@@ -0,0 +1,356 @@
+            ],
+        ]);
+    }
+    
+    protected function loadValuesFromConfig($config, $attributeRules)
+    {
+        foreach ($attributeRules as $attribute => $rules) {
+            
+            if (array_key_exists($attribute, $config)) {
+                $this->$attribute = $config[$attribute];
+            }
+            
+            Validator::validate($this->$attribute, $rules, $this->logger, $attribute);
+        }
+    }
+    
+    /**
+     * Get the specified attribute from the given state data.
+     *
+     * NOTE: If the attribute's data is an array, the first value will be
+     *       returned. Otherwise, the attribute's data will simply be returned
+     *       as-is.
+     *
+     * @param string $attributeName The name of the attribute.
+     * @param array $state The state data.
+     * @return mixed The attribute value, or null if not found.
+     */
+    protected function getAttribute($attributeName, $state)
+    {
+        $attributeData = $state['Attributes'][$attributeName] ?? null;
+        
+        if (is_array($attributeData)) {
+            return $attributeData[0] ?? null;
+        }
+        
+        return $attributeData;
+    }
+    
+    /**
+     * Calculate how many days remain between now and when the password will
+     * expire.
+     *
+     * @param int $expiryTimestamp The timestamp for when the password will
+     *     expire.
+     * @return int The number of days remaining
+     */
+    protected function getDaysLeftBeforeExpiry($expiryTimestamp)
+    {
+        $now = time();
+        $end = $expiryTimestamp;
+        return round(($end - $now) / (24*60*60));
+    }
+    
+    /**
+     * Get the timestamp for when the user's password will expire, throwing an
+     * exception if unable to do so.
+     *
+     * @param string $expiryDateAttr The name of the attribute where the
+     *     expiration date (as a string) is stored.
+     * @param array $state The state data.
+     * @return int The expiration timestamp.
+     * @throws Exception
+     */
+    protected function getExpiryTimestamp($expiryDateAttr, $state)
+    {
+        $expiryDateString = $this->getAttribute($expiryDateAttr, $state);
+        
+        // Ensure that EVERY user login provides a usable password expiration date.
+        $expiryTimestamp = strtotime($expiryDateString) ?: null;
+        if (empty($expiryTimestamp)) {
+            throw new Exception(sprintf(
+                "We could not understand the expiration date (%s, from %s) for "
+                . "the user's password, so we do not know whether their "
+                . "password is still valid.",
+                var_export($expiryDateString, true),
+                var_export($expiryDateAttr, true)
+            ), 1496843359);
+        }
+        return $expiryTimestamp;
+    }
+    
+    protected function initComposerAutoloader()
+    {
+        $path = __DIR__ . '/../../../vendor/autoload.php';
+        if (file_exists($path)) {
```

Comment:
So if this path is incorrect it will not load the file and will not throw an exception, is that okay? I'm not sure why the path would change but if it doesnt load the autoloader we'll just get other errors about class not defined stuff.